frontend/src/components/common/content/ContentBlockRenderer.ts

```diff
@@ -0,0 +1,64 @@
+import React from "react";
+import { ContentType } from "../../../APIClients/types/LessonClientTypes";
+import { ContentBlockState } from "../../../types/ContentBlockTypes";
+
+// Many thanks to https://github.com/Microsoft/TypeScript/issues/30581#issuecomment-1080979994.
+
+type ValueOf<T> = T[keyof T];
+
+type ArgMap<D extends ContentType> = {
+  [K in D]: Extract<ContentBlockState, { ["type"]: { ["clientType"]: K } }>;
+};
+
+class ContentBlockRenderer<
+  PropType extends { block: ContentBlockState },
+  DispatchableT extends ContentType = ContentType
+> {
+  constructor(
+    private dispatchTable: {
+      [K in keyof ArgMap<DispatchableT>]: (
+        props: Omit<PropType, "block"> & { block: ArgMap<DispatchableT>[K] },
+      ) => React.ReactNode;
+    },
+  ) {}
+
+  private callWithMethodAndArgs(
+    op: {
+      [P in keyof ArgMap<DispatchableT>]: {
+        method: P;
+        arg: Omit<PropType, "block"> & { block: ArgMap<DispatchableT>[P] };
+      };
+    }[keyof ArgMap<DispatchableT>],
+  ) {
+    return this.dispatchTable[op.method](op.arg);
+  }
+
+  private callWithArgMap(
+    props: Omit<PropType, "block"> & { block: ValueOf<ArgMap<DispatchableT>> },
+  ) {
+    return this.callWithMethodAndArgs({
+      method: props.block.type.clientType,
+      arg: props,
+    });
+  }
+
+  public canDispatch(
+    block: ContentBlockState,
+  ): block is ValueOf<ArgMap<DispatchableT>> {
+    return block.type.clientType in this.dispatchTable;
+  }
+
+  public render(props: PropType): React.ReactNode {
+    const { block, ...rest } = props;
+
+    if (this.canDispatch(block)) {
+      return this.callWithArgMap({ block, ...rest });
+    }
```

Comment:
Can this not be written more simply? I get that this was copied from the github issue, but it could be improved :)
```javascript
const { block } = props;

if (this.canDispatch(block)) {
   return this.callWithArgMap(props);
}
```

Reply:
Unfortunately, no, actually! Or at least, I wasn't able to figure out how to make it work. If you implement the solution as given above, VSCode complains:
![image](https://user-images.githubusercontent.com/9922514/175460885-8a05ff5c-bc6b-47db-9c99-ea6ee53314d0.png)

I think it's because with the current solution, Typescript knows that the type of `block` was narrowed, but it's not smart enough to realize that the `block` which we extracted is the same `block` as is contained in `props`. I could just cast it and avoid the extra verbosity, but I wanted to avoid casts here if possible.

Reply:
Dang, seems like sketchy typescript stuff to me. Good to know - I'm happy with it then :)